src/files.c

```diff
@@ -785,32 +785,48 @@ bool verify_FTL(bool verbose)
 	if(readlink("/proc/self/exe", filename, sizeof(filename)) == -1)
 	{
 		log_err("Failed to read self filename: %s", strerror(errno));
-		return -1;
+		return false;
 	}
 
-	// Read the pre-computed hash - it is stored in the last 8 bytes of the
-	// binary itself
+	// Read the pre-computed hash as well as the checksum mark from the
+	// binary (last 9 + 32 bytes)
 	uint8_t self_hash[SHA256_DIGEST_SIZE];
+	uint8_t checksum_mark[9] = { 0 };
 	FILE *f = fopen(filename, "r");
 	if(f == NULL)
 	{
 		log_err("Failed to open self file \"%s\": %s", filename, strerror(errno));
-		return -1;
+		return false;
 	}
-	if(fseek(f, -SHA256_DIGEST_SIZE, SEEK_END) != 0)
+	if(fseek(f, -(SHA256_DIGEST_SIZE + 9), SEEK_END) != 0)
 	{
 		log_err("Failed to seek to hash: %s", strerror(errno));
 		fclose(f);
-		return -1;
+		return false;
+	}
+	if(fread(checksum_mark, 9, 1, f) != 1)
+	{
+		log_err("Failed to read checksum mark: %s", strerror(errno));
+		fclose(f);
+		return false;
 	}
 	if(fread(self_hash, SHA256_DIGEST_SIZE, 1, f) != 1)
 	{
 		log_err("Failed to read hash: %s", strerror(errno));
 		fclose(f);
-		return -1;
+		return false;
 	}
 	fclose(f);
 
+	// Verify checksum mark matches magic string "sha256sum"
+	if(memcmp(checksum_mark, "sha256sum", 9) != 0)
+	{
+		log_warn("Binary integrity check not possible: No checksum mark found");
+		// This is not an error, as the binary may not have a checksum mark
+		// if it was built with a different toolchain
+		return true;
```

Comment:
Should we really return `true`here? This is an undefined state, where we can neither verify nor reject the integrity.  We would need a third state. If returning `true` here, the CLI will report binary is Ok and only the log will print the warning.

Reply:
Hmm, yes, I agree. Looking into `enum.h` it really doesn't look like we have an aversion against adding `enums` when we need them, so let's add another here:
``` c
enum verify_result {
	VERIFY_OK = 0, // EXIT_SUCCESS
	VERIFY_FAILED, // 1
	VERIFY_ERROR, // 2
	VERIFY_NO_CHECKSUM // 3
} __attribute__ ((packed));
```
which is directly exposed as exit code. As everything `!= 0` is `false` in (ba)sh, this preserves the previous behavior while still exposing more detail to the outside world. I am wondering if we need/should document this somewhere but after looking at `pihole-FTL -h`, this really would go beyond the scope of this "document".

Reply:
I seldom admit it, but more documentation is not necessary I think.